### PR TITLE
Ensure correct vTables when interface class is redefined

### PR DESCRIPTION
**Pull request by gacholio**
Ensure that all classes that implement a redefined interface have their
vTable methods updated to the new methods. Add a test to verify.

[ci skip]

Signed-off-by: Graham Chapman <graham_chapman@ca.ibm.com>

### CODE REVIEW

**Comment by gacholio** [comment]
@tajila

**Comment by gacholio** [comment]
jenkins test sanity,extended zlinux jdk8

File: runtime/util/hshelp.c [hunk 1]
```diff
@@ -2701,59 +2701,62 @@ addClassesRequiringNewITables(J9JavaVM *vm, J9HashTable *classHashTable, UDATA *
...
-
-								/* If this class is already in the table this add will have no effect */
-								result = hashTableAdd(classHashTable, &exemplar);
-								if (NULL == result) {
-									vmFuncs->allClassesEndDo(&classWalkState);
-									return JVMTI_ERROR_OUT_OF_MEMORY;
-								}
-								/* Ensure that classes already in the table get redefined */
-								result->flags |= J9JVMTI_CLASS_PAIR_FLAG_REDEFINED;
+					/* An interface of this class is being updated with extensions - this class will require new iTables. */
```
Comment (DanHeidinga) [inline runtime/util/hshelp.c:2720]: Is this comment still accurate?  Maybe it needs to move up and into the `if (!fastHCR) {` block?

File: runtime/util/hshelp.c [hunk 2]
```diff
@@ -2864,20 +2869,16 @@ determineClassesToRecreate(J9VMThread * currentThread, jint classCount,
...
-		/* Add any classes implementing interfaces that are being updated with extensions or re-ordered methods. */
-		jvmtiResult = addClassesRequiringNewITables(currentThread->javaVM, classHashTable, &addedClassCount, &addedMethodCount);
-		if (JVMTI_ERROR_NONE != jvmtiResult) {
-			hashTableFree(classHashTable);
-			return jvmtiResult;
-		}
-
-		redefinedMethodCount += addedMethodCount;
-		classCount += (jint) addedClassCount;
+	/* Add any classes implementing interfaces that are being updated with extensions or re-ordered methods. */
```
Comment (DanHeidinga) [inline runtime/util/hshelp.c:2872]: Comment still apply?

File: runtime/tests/jvmtitests/agent/tests.c [hunk 3]
```diff
@@ -95,6 +95,7 @@ static jvmtiTest jvmtiTestList[] =
 	{ "rc019a", rc019a, "com.ibm.jvmti.tests.redefineClasses.rc019a", "RedefineClasses" },
 	{ "rc019b", rc019b, "com.ibm.jvmti.tests.redefineClasses.rc019b", "RedefineClasses" },
 	{ "rc020", rc020, "com.ibm.jvmti.tests.redefineClasses.rc020", "RedefineClasses" },
+	{ "rc021", rc021, "com.ibm.jvmti.tests.redefineClasses.rc021", "RedefineClasses" },
```
Comment (DanHeidinga) [inline runtime/tests/jvmtitests/agent/tests.c:98]:
```suggestion
	{ "rc021", rc021, "com.ibm.jvmti.tests.redefineClasses.rc021", "RetransformClasses" },
```

**Review by DanHeidinga — COMMENTED** [review]
Does this apply even when extensions aren't used?  Does it relate to defender methods?  Not sure I follow why this change is necessary

**Comment by tajila** [comment]
> Does this apply even when extensions aren't used? Does it relate to defender methods? Not sure I follow why this change is necessary

Yes, this relates to default methods. Currently when we modify a classes methods in fastHCR, we fixup all the subclasses that would have the supers methods in the vtable. We need to do that for implementing classes as well (currenly only done in extensions) has they might have obsolete defaults methods in their vtable.

File: runtime/util/hshelp.c (hunk 1 above)
Comment (gacholio) [reply runtime/util/hshelp.c]: Moved up.

File: runtime/util/hshelp.c (hunk 2 above)
Comment (gacholio) [reply runtime/util/hshelp.c:2872]: Yes.

File: runtime/tests/jvmtitests/agent/tests.c (hunk 3 above)
Comment (gacholio) [reply runtime/tests/jvmtitests/agent/tests.c]: Renamed test to rtc003 - the distinction is strictly notational. Retransform is just a much easier way to do a NOP HCR than trying to find the class bytes.

**Comment by gacholio** [comment]
jenkins test sanity,extended zlinux jdk8

**Review by DanHeidinga — APPROVED** [review]
looks ok to me.  @tajila ?

**Comment by gacholio** [comment]
jenkins test sanity,extended plinux jdk11

**Review by tajila — APPROVED** [review]

**Pull request merged**